# Notebook 04c — Temperature Forecast with Hydrological Inputs

Thiago Nascimento (Eawag) suggested using precipitation as an input feature
for temperature forecasting. Catchment-scale precipitation data from CAMELS-CH
base forcing (Höge et al. 2023) is not available locally, so we use electrical
conductivity (EC) as a precipitation proxy — EC drops sharply during rain events
due to dilution of dissolved minerals (negative correlation with precip, r ≈ −0.4
in Alpine catchments). This notebook:
1. Trains a temperature LSTM using EC as an additional input (proxy for dilution/precip)
2. Compares to the baseline temperature model (temp only)
3. Evaluates across all 86 gauges that have both temp and EC data
4. Notes the framework for integrating actual CAMELS-CH precipitation forcing when available


In [1]:
import sys, warnings
from pathlib import Path
import numpy as np
import pandas as pd
import torch
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import seaborn as sns
from tqdm.auto import tqdm
from sklearn.preprocessing import StandardScaler

warnings.filterwarnings('ignore')
sys.path.insert(0, '..')

from src.config import (
    LOOKBACK, HORIZON, TRAIN_END, VAL_END, SEED,
    FOCUS_GAUGE, RESULTS_DIR, FIGURES_DIR, DATA_ROOT,
)
from src.data import load_gauge, preprocess, train_val_test_split, make_windows
from src.model import (
    Seq2SeqLSTM, RiverDataset, train_model, predict, get_y_true,
)
from src.metrics import mean_rmse, nse, kge

print(f'GPU: {torch.cuda.get_device_name(0)}')

sns.set_style('whitegrid')
plt.rcParams.update({'figure.dpi': 130})

# Features: temp-only vs temp+EC

DATA_DIR = DATA_ROOT / 'stream_water_chemistry/timeseries/daily'

# LOCAL_TEST: True = fast CPU verification, False = full run
LOCAL_TEST = False  # set True for quick CPU verification
if LOCAL_TEST:
    print('LOCAL_TEST mode: reduced epochs for quick verification')

np.random.seed(SEED)
torch.manual_seed(SEED)
DEVICE = 'cuda' if torch.cuda.is_available() else 'cpu'
print(f'Device: {DEVICE}')

# Model hyperparameters
HIDDEN, N_LAYERS, DROPOUT = 64, 2, 0.2
LR, EPOCHS, PATIENCE = 1e-3, 50, 8
EPOCHS_MULTI, PATIENCE_MULTI = 30, 5   # lighter for multi-gauge loop

# Feature sets
FEATURES_TEMP_ONLY = ['temp_sensor']
FEATURES_TEMP_EC   = ['temp_sensor', 'ec_sensor']
TARGET_TEMP        = ['temp_sensor']

# Data directory
REPO_ROOT = Path('/storage/homefs/tn20y076/AareML')


GPU: NVIDIA GeForce RTX 4090
Device: cuda


## 1. EC as Precipitation Proxy

Electrical conductivity (EC) decreases during rain events due to dilution of
dissolved minerals — this is a well-established relationship in hydrology
(Godsey et al. 2009, *Hydrological Processes*). In Alpine catchments, the
negative EC–precipitation correlation is particularly strong (r ≈ −0.4 to −0.7)
because of rapid snowmelt and storm-driven runoff.

EC is available at all CAMELS-CH-Chem gauges (`ec_sensor` column), making it
a practical proxy until the CAMELS-CH meteorological forcing data is linked.


In [2]:
raw  = load_gauge(FOCUS_GAUGE)
data = preprocess(raw)

# Overall EC–temperature correlation at focus gauge
corr = data[['ec_sensor', 'temp_sensor']].corr().loc['ec_sensor', 'temp_sensor']
print(f'EC–temperature correlation at gauge {FOCUS_GAUGE}: r = {corr:.3f}')

# Rolling 90-day correlation to show seasonal structure
roll_corr = data['ec_sensor'].rolling(90).corr(data['temp_sensor'])

fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(12, 6), sharex=True)

ax1.plot(data.index, data['temp_sensor'], color='#d62728', alpha=0.8,
         label='Temperature (°C)', linewidth=0.8)
ax1_r = ax1.twinx()
ax1_r.plot(data.index, data['ec_sensor'], color='#2ca02c', alpha=0.5,
           label='EC (µS/cm)', linewidth=0.6)
ax1.set_ylabel('Temperature (°C)', color='#d62728')
ax1_r.set_ylabel('EC (µS/cm)', color='#2ca02c')
ax1.set_title(f'Temperature and EC at Gauge {FOCUS_GAUGE} (Aare at Bern)')
lines1, labels1 = ax1.get_legend_handles_labels()
lines2, labels2 = ax1_r.get_legend_handles_labels()
ax1.legend(lines1 + lines2, labels1 + labels2, loc='upper right', fontsize=8)

ax2.plot(data.index, roll_corr, color='#9467bd', linewidth=1.2,
         label='90-day rolling r(EC, Temp)')
ax2.axhline(0, color='black', linestyle='--', linewidth=0.8)
ax2.fill_between(data.index, roll_corr, 0,
                 where=(roll_corr < 0), alpha=0.2, color='#d62728',
                 label='Negative correlation (EC ↓ when Temp ↑)')
ax2.set_ylabel('90-day rolling r(EC, Temp)')
ax2.set_xlabel('Date')
ax2.legend(fontsize=8)

plt.tight_layout()
plt.savefig(FIGURES_DIR / 'nb04c_ec_temp_correlation.png', dpi=150, bbox_inches='tight')
plt.close()
print('Figure saved: nb04c_ec_temp_correlation.png')


[data] load_gauge 2473: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.0, 'pH_sensor': 0.01, 'ec_sensor': 0.007, 'O2C_sensor': 0.012}
EC–temperature correlation at gauge 2473: r = -0.549


Figure saved: nb04c_ec_temp_correlation.png


## 2. Single-Site LSTM: Temp + EC vs Temp-Only

Train two Seq2Seq LSTM models on the focus gauge (2473 — Aare at Bern):
- **Temp-only baseline**: autoregressive, uses only `temp_sensor`
- **Temp + EC**: adds `ec_sensor` as second input feature

Same architecture for both: hidden=64, n_layers=2, dropout=0.2.


In [3]:
from sklearn.preprocessing import StandardScaler

def build_and_train(features, label):
    raw  = load_gauge(FOCUS_GAUGE)
    data = preprocess(raw)
    train_df, val_df, test_df = train_val_test_split(data)
    train_means = (pd.concat([train_df[features].mean(), train_df[TARGET_TEMP].mean()])
                   .groupby(level=0).first())

    X_tr, y_tr, _ = make_windows(train_df, train_means, features=features, targets=TARGET_TEMP)
    X_va, y_va, _ = make_windows(val_df,   train_means, features=features, targets=TARGET_TEMP)
    X_te, y_te, _ = make_windows(test_df,  train_means, features=features, targets=TARGET_TEMP)

    N_tr, L, F = X_tr.shape
    H, T = y_tr.shape[1], y_tr.shape[2]

    feat_sc = StandardScaler().fit(X_tr.reshape(-1, F))
    tgt_sc  = StandardScaler().fit(y_tr.reshape(-1, T))

    def scale(X_raw, y_raw, N, _F=F, _L=L, _T=T, _H=H, _fsc=feat_sc, _tsc=tgt_sc):
        Xs = _fsc.transform(X_raw.reshape(-1, _F)).reshape(N, _L, _F).astype('float32')
        ys = _tsc.transform(y_raw.reshape(-1, _T)).reshape(N, _H, _T).astype('float32')
        return Xs, ys

    Xs_tr, ys_tr = scale(X_tr, y_tr, N_tr)
    Xs_va, ys_va = scale(X_va, y_va, X_va.shape[0])
    Xs_te, ys_te = scale(X_te, y_te, X_te.shape[0])

    ds_tr = RiverDataset(Xs_tr, ys_tr)
    ds_va = RiverDataset(Xs_va, ys_va)
    ds_te = RiverDataset(Xs_te, ys_te)

    dl_tr = torch.utils.data.DataLoader(ds_tr, batch_size=256, shuffle=True)
    dl_va = torch.utils.data.DataLoader(ds_va, batch_size=256, shuffle=False)

    model = Seq2SeqLSTM(n_feat=F, n_tgt=T,
                         hidden=HIDDEN, n_layers=N_LAYERS, dropout=DROPOUT).to(DEVICE)
    model, _ = train_model(model, dl_tr, dl_va, device=DEVICE,
                            epochs=EPOCHS, lr=LR, patience=PATIENCE, verbose=False)

    # predict() handles DataLoader + inverse_transform internally
    y_pred = predict(model, ds_te, tgt_sc, device=DEVICE)   # shape (N, H, T)
    y_true = get_y_true(ds_te, tgt_sc)                       # shape (N, H, T)

    rmse = float(np.sqrt(np.mean((y_pred - y_true)**2)))
    ss_res = np.sum((y_true - y_pred)**2)
    ss_tot = np.sum((y_true - y_true.mean())**2)
    nse_ = float(1 - ss_res/ss_tot) if ss_tot > 0 else float('nan')

    print(f'  [{label}] test RMSE={rmse:.3f}°C  NSE={nse_:.3f}')
    return model, feat_sc, tgt_sc, {'rmse': rmse, 'nse': nse_}

print('Training temperature-only model...')
model_base, _, tgt_sc_base, metrics_base = build_and_train(FEATURES_TEMP_ONLY, 'temp-only')

print('Training temperature + EC model...')
model_ec, _, tgt_sc_ec, metrics_ec = build_and_train(FEATURES_TEMP_EC, 'temp+EC')

print()
print('=' * 50)
print(f'Single-site comparison at gauge {FOCUS_GAUGE}:')
print(f'  Temp-only:  RMSE={metrics_base["rmse"]:.3f}°C  NSE={metrics_base["nse"]:.3f}')
print(f'  Temp + EC:  RMSE={metrics_ec["rmse"]:.3f}°C  NSE={metrics_ec["nse"]:.3f}')
delta = metrics_base['rmse'] - metrics_ec['rmse']
print(f'  EC improvement: {delta:+.3f}°C ({delta/metrics_base["rmse"]*100:+.1f}%)')


Training temperature-only model...
[data] load_gauge 2473: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.0, 'pH_sensor': 0.01, 'ec_sensor': 0.007, 'O2C_sensor': 0.012}
[data] split: train=12418, val=731, test=1461
[data] make_windows: 12384 windows, X=(12384, 21, 1), y=(12384, 14, 1), date range 1981-01-22 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 1), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1427 windows, X=(1427, 21, 1), y=(1427, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 12384 samples, X=(12384, 21, 1), y=(12384, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 1), y=(697, 14, 1)
[model] RiverDataset: 1427 samples, X=(1427, 21, 1), y=(1427, 14, 1)


[model] train_model: 100,929 trainable params, device=cuda, epochs=50, patience=8, lr=0.001


[model] predict: 1427 samples, DO range [-1.74, 2.59] mg/L (scaled)
  [temp-only] test RMSE=1.314°C  NSE=0.875
Training temperature + EC model...
[data] load_gauge 2473: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.0, 'pH_sensor': 0.01, 'ec_sensor': 0.007, 'O2C_sensor': 0.012}
[data] split: train=12418, val=731, test=1461
[data] make_windows: 12384 windows, X=(12384, 21, 2), y=(12384, 14, 1), date range 1981-01-22 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 2), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1427 windows, X=(1427, 21, 2), y=(1427, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 12384 samples, X=(12384, 21, 2), y=(12384, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 2), y=(697, 14, 1)
[model] RiverDataset: 1427 samples, X=(1427, 21, 2), y=(1427, 14, 1)
[model] train_model: 101,185 trainable params, dev

[model] predict: 1427 samples, DO range [-1.75, 2.59] mg/L (scaled)
  [temp+EC] test RMSE=1.240°C  NSE=0.889

Single-site comparison at gauge 2473:
  Temp-only:  RMSE=1.314°C  NSE=0.875
  Temp + EC:  RMSE=1.240°C  NSE=0.889
  EC improvement: +0.074°C (+5.7%)


## 3. Multi-Gauge Evaluation: Temp+EC vs Temp-Only

Evaluate both models across all gauges with both `temp_sensor` **and**
`ec_sensor` data, each with >365 days of valid readings.
For each gauge we train from scratch and evaluate on the held-out test period.


In [4]:
print('Scanning gauges with temp + EC data...')
valid_gauges = []
data_dir = REPO_ROOT / 'data/camels-ch-chem/stream_water_chemistry/timeseries/daily'
for f in sorted(data_dir.glob('camels_ch_chem_daily_*.csv')):
    gid = int(f.stem.split('_')[-1])
    if str(gid) == str(FOCUS_GAUGE):
        continue
    try:
        df = pd.read_csv(f, parse_dates=['date'])
        n_temp = df['temp_sensor'].notna().sum() if 'temp_sensor' in df.columns else 0
        n_ec   = df['ec_sensor'].notna().sum()   if 'ec_sensor'   in df.columns else 0
        # Require both in training AND test period
        df['date'] = pd.to_datetime(df['date'])
        test_df_raw = df[df['date'] > pd.Timestamp(VAL_END)]
        n_temp_test = test_df_raw['temp_sensor'].notna().sum() if 'temp_sensor' in df.columns else 0
        n_ec_test   = test_df_raw['ec_sensor'].notna().sum()   if 'ec_sensor'   in df.columns else 0
        if n_temp >= 365 and n_ec >= 365 and n_temp_test >= 50 and n_ec_test >= 50:
            valid_gauges.append(gid)
    except:
        pass
print(f'Gauges with temp+EC: {len(valid_gauges)}')

all_results = []
for gid in tqdm(valid_gauges, desc='Multi-gauge eval'):
    for features, label in [(FEATURES_TEMP_ONLY, 'base'), (FEATURES_TEMP_EC, 'ec')]:
        try:
            raw  = load_gauge(gid)
            data = preprocess(raw)
            train_df, val_df, test_df = train_val_test_split(data)
            train_means = (pd.concat([train_df[features].mean(), train_df[TARGET_TEMP].mean()])
                           .groupby(level=0).first())

            X_tr, y_tr, _ = make_windows(train_df, train_means, features=features, targets=TARGET_TEMP)
            X_va, y_va, _ = make_windows(val_df,   train_means, features=features, targets=TARGET_TEMP)
            X_te, y_te, _ = make_windows(test_df,  train_means, features=features, targets=TARGET_TEMP)
            if len(X_te) < 10: continue

            N_tr, L, F = X_tr.shape
            H, T = y_tr.shape[1], y_tr.shape[2]
            feat_sc = StandardScaler().fit(X_tr.reshape(-1, F))
            tgt_sc  = StandardScaler().fit(y_tr.reshape(-1, T))

            def _scale(X_raw, y_raw, N, _F=F, _L=L, _T=T, _H=H, _fsc=feat_sc, _tsc=tgt_sc):
                Xs = _fsc.transform(X_raw.reshape(-1, _F)).reshape(N, _L, _F).astype('float32')
                ys = _tsc.transform(y_raw.reshape(-1, _T)).reshape(N, _H, _T).astype('float32')
                return Xs, ys

            Xs_tr, ys_tr = _scale(X_tr, y_tr, N_tr)
            Xs_va, ys_va = _scale(X_va, y_va, X_va.shape[0])
            Xs_te, ys_te = _scale(X_te, y_te, X_te.shape[0])

            ds_tr = RiverDataset(Xs_tr, ys_tr)
            ds_va = RiverDataset(Xs_va, ys_va)
            ds_te = RiverDataset(Xs_te, ys_te)

            dl_tr = torch.utils.data.DataLoader(ds_tr, batch_size=256, shuffle=True)
            dl_va = torch.utils.data.DataLoader(ds_va, batch_size=256, shuffle=False)

            mdl = Seq2SeqLSTM(n_feat=F, n_tgt=T, hidden=HIDDEN,
                               n_layers=N_LAYERS, dropout=DROPOUT).to(DEVICE)
            mdl, _ = train_model(mdl, dl_tr, dl_va, device=DEVICE,
                                  epochs=EPOCHS_MULTI, lr=LR, patience=PATIENCE_MULTI, verbose=False)

            y_pred = predict(mdl, ds_te, tgt_sc, device=DEVICE)
            y_true = get_y_true(ds_te, tgt_sc)

            rmse_v = float(np.sqrt(np.mean((y_pred - y_true)**2)))
            ss_res = np.sum((y_true - y_pred)**2)
            ss_tot = np.sum((y_true - y_true.mean())**2)
            nse_v  = float(1 - ss_res/ss_tot) if ss_tot > 0 else float('nan')

            all_results.append({'gauge_id': gid, 'model': label,
                                 'rmse_temp': round(rmse_v, 4), 'nse_temp': round(nse_v, 3)})
        except Exception as e:
            pass

df_mg = pd.DataFrame(all_results)
if df_mg.empty:
    print("No results collected")
else:
    base_r = df_mg[df_mg.model=='base']
    ec_r   = df_mg[df_mg.model=='ec']
    print(f'Multi-gauge results ({len(base_r)} gauges):')
    print(f'  Temp-only: mean RMSE={base_r.rmse_temp.mean():.3f}°C  NSE={base_r.nse_temp.mean():.3f}')
    print(f'  Temp+EC:   mean RMSE={ec_r.rmse_temp.mean():.3f}°C  NSE={ec_r.nse_temp.mean():.3f}')
    df_mg.to_csv(RESULTS_DIR / 'temp_ec_results.csv', index=False)
    print(f'Saved: temp_ec_results.csv')


Scanning gauges with temp + EC data...


Gauges with temp+EC: 14


Multi-gauge eval:   0%|          | 0/14 [00:00<?, ?it/s]

[data] load_gauge 2009: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.0, 'pH_sensor': 0.022, 'ec_sensor': 0.011, 'O2C_sensor': 0.016}
[data] split: train=12418, val=731, test=1461
[data] make_windows: 12384 windows, X=(12384, 21, 1), y=(12384, 14, 1), date range 1981-01-22 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 1), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1427 windows, X=(1427, 21, 1), y=(1427, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 12384 samples, X=(12384, 21, 1), y=(12384, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 1), y=(697, 14, 1)
[model] RiverDataset: 1427 samples, X=(1427, 21, 1), y=(1427, 14, 1)
[model] train_model: 100,929 trainable params, device=cuda, epochs=30, patience=5, lr=0.001


[model] predict: 1427 samples, DO range [-1.64, 1.72] mg/L (scaled)
[data] load_gauge 2009: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.0, 'pH_sensor': 0.022, 'ec_sensor': 0.011, 'O2C_sensor': 0.016}
[data] split: train=12418, val=731, test=1461
[data] make_windows: 12384 windows, X=(12384, 21, 2), y=(12384, 14, 1), date range 1981-01-22 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 2), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1427 windows, X=(1427, 21, 2), y=(1427, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 12384 samples, X=(12384, 21, 2), y=(12384, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 2), y=(697, 14, 1)
[model] RiverDataset: 1427 samples, X=(1427, 21, 2), y=(1427, 14, 1)
[model] train_model: 101,185 trainable params, device=cuda, epochs=30, patience=5, lr=0.001


[model] predict: 1427 samples, DO range [-1.63, 1.60] mg/L (scaled)
[data] load_gauge 2016: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.0, 'pH_sensor': 0.054, 'ec_sensor': 0.05, 'O2C_sensor': 0.063}
[data] split: train=12418, val=731, test=1461


[data] make_windows: 12384 windows, X=(12384, 21, 1), y=(12384, 14, 1), date range 1981-01-22 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 1), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1427 windows, X=(1427, 21, 1), y=(1427, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 12384 samples, X=(12384, 21, 1), y=(12384, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 1), y=(697, 14, 1)
[model] RiverDataset: 1427 samples, X=(1427, 21, 1), y=(1427, 14, 1)
[model] train_model: 100,929 trainable params, device=cuda, epochs=30, patience=5, lr=0.001


[model] predict: 1427 samples, DO range [-1.57, 2.09] mg/L (scaled)
[data] load_gauge 2016: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.0, 'pH_sensor': 0.054, 'ec_sensor': 0.05, 'O2C_sensor': 0.063}
[data] split: train=12418, val=731, test=1461
[data] make_windows: 12384 windows, X=(12384, 21, 2), y=(12384, 14, 1), date range 1981-01-22 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 2), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1427 windows, X=(1427, 21, 2), y=(1427, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 12384 samples, X=(12384, 21, 2), y=(12384, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 2), y=(697, 14, 1)
[model] RiverDataset: 1427 samples, X=(1427, 21, 2), y=(1427, 14, 1)
[model] train_model: 101,185 trainable params, device=cuda, epochs=30, patience=5, lr=0.001


[model] predict: 1427 samples, DO range [-1.57, 2.08] mg/L (scaled)
[data] load_gauge 2018: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.0, 'pH_sensor': 0.576, 'ec_sensor': 0.564, 'O2C_sensor': 0.568}
[data] split: train=12418, val=731, test=1461
[data] make_windows: 12384 windows, X=(12384, 21, 1), y=(12384, 14, 1), date range 1981-01-22 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 1), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1427 windows, X=(1427, 21, 1), y=(1427, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 12384 samples, X=(12384, 21, 1), y=(12384, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 1), y=(697, 14, 1)
[model] RiverDataset: 1427 samples, X=(1427, 21, 1), y=(1427, 14, 1)
[model] train_model: 100,929 trainable params, device=cuda, epochs=30, patience=5, lr=0.001


[model] predict: 1427 samples, DO range [-1.50, 2.21] mg/L (scaled)
[data] load_gauge 2018: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.0, 'pH_sensor': 0.576, 'ec_sensor': 0.564, 'O2C_sensor': 0.568}
[data] split: train=12418, val=731, test=1461


[data] make_windows: 12384 windows, X=(12384, 21, 2), y=(12384, 14, 1), date range 1981-01-22 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 2), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1427 windows, X=(1427, 21, 2), y=(1427, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 12384 samples, X=(12384, 21, 2), y=(12384, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 2), y=(697, 14, 1)
[model] RiverDataset: 1427 samples, X=(1427, 21, 2), y=(1427, 14, 1)
[model] train_model: 101,185 trainable params, device=cuda, epochs=30, patience=5, lr=0.001


[model] predict: 1427 samples, DO range [-1.50, 2.08] mg/L (scaled)
[data] load_gauge 2044: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.0, 'pH_sensor': 0.196, 'ec_sensor': 0.193, 'O2C_sensor': 0.213}
[data] split: train=12418, val=731, test=1461
[data] make_windows: 12384 windows, X=(12384, 21, 1), y=(12384, 14, 1), date range 1981-01-22 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 1), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1427 windows, X=(1427, 21, 1), y=(1427, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 12384 samples, X=(12384, 21, 1), y=(12384, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 1), y=(697, 14, 1)
[model] RiverDataset: 1427 samples, X=(1427, 21, 1), y=(1427, 14, 1)
[model] train_model: 100,929 trainable params, device=cuda, epochs=30, patience=5, lr=0.001


[model] predict: 1427 samples, DO range [-1.43, 1.89] mg/L (scaled)
[data] load_gauge 2044: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.0, 'pH_sensor': 0.196, 'ec_sensor': 0.193, 'O2C_sensor': 0.213}
[data] split: train=12418, val=731, test=1461
[data] make_windows: 12384 windows, X=(12384, 21, 2), y=(12384, 14, 1), date range 1981-01-22 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 2), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1427 windows, X=(1427, 21, 2), y=(1427, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 12384 samples, X=(12384, 21, 2), y=(12384, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 2), y=(697, 14, 1)
[model] RiverDataset: 1427 samples, X=(1427, 21, 2), y=(1427, 14, 1)
[model] train_model: 101,185 trainable params, device=cuda, epochs=30, patience=5, lr=0.001


[model] predict: 1427 samples, DO range [-1.49, 2.14] mg/L (scaled)
[data] load_gauge 2085: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.0, 'pH_sensor': 0.455, 'ec_sensor': 0.357, 'O2C_sensor': 0.463}
[data] split: train=12418, val=731, test=1461
[data] make_windows: 12384 windows, X=(12384, 21, 1), y=(12384, 14, 1), date range 1981-01-22 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 1), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1427 windows, X=(1427, 21, 1), y=(1427, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 12384 samples, X=(12384, 21, 1), y=(12384, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 1), y=(697, 14, 1)
[model] RiverDataset: 1427 samples, X=(1427, 21, 1), y=(1427, 14, 1)
[model] train_model: 100,929 trainable params, device=cuda, epochs=30, patience=5, lr=0.001


[model] predict: 1427 samples, DO range [-1.60, 2.03] mg/L (scaled)
[data] load_gauge 2085: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.0, 'pH_sensor': 0.455, 'ec_sensor': 0.357, 'O2C_sensor': 0.463}
[data] split: train=12418, val=731, test=1461
[data] make_windows: 12384 windows, X=(12384, 21, 2), y=(12384, 14, 1), date range 1981-01-22 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 2), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1427 windows, X=(1427, 21, 2), y=(1427, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 12384 samples, X=(12384, 21, 2), y=(12384, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 2), y=(697, 14, 1)
[model] RiverDataset: 1427 samples, X=(1427, 21, 2), y=(1427, 14, 1)
[model] train_model: 101,185 trainable params, device=cuda, epochs=30, patience=5, lr=0.001


[model] predict: 1427 samples, DO range [-1.69, 2.34] mg/L (scaled)
[data] load_gauge 2143: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.0, 'pH_sensor': 0.022, 'ec_sensor': 0.012, 'O2C_sensor': 0.016}
[data] split: train=12418, val=731, test=1461
[data] make_windows: 12384 windows, X=(12384, 21, 1), y=(12384, 14, 1), date range 1981-01-22 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 1), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1427 windows, X=(1427, 21, 1), y=(1427, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 12384 samples, X=(12384, 21, 1), y=(12384, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 1), y=(697, 14, 1)
[model] RiverDataset: 1427 samples, X=(1427, 21, 1), y=(1427, 14, 1)
[model] train_model: 100,929 trainable params, device=cuda, epochs=30, patience=5, lr=0.001


[model] predict: 1427 samples, DO range [-1.56, 2.38] mg/L (scaled)
[data] load_gauge 2143: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.0, 'pH_sensor': 0.022, 'ec_sensor': 0.012, 'O2C_sensor': 0.016}
[data] split: train=12418, val=731, test=1461
[data] make_windows: 12384 windows, X=(12384, 21, 2), y=(12384, 14, 1), date range 1981-01-22 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 2), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1427 windows, X=(1427, 21, 2), y=(1427, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 12384 samples, X=(12384, 21, 2), y=(12384, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 2), y=(697, 14, 1)
[model] RiverDataset: 1427 samples, X=(1427, 21, 2), y=(1427, 14, 1)
[model] train_model: 101,185 trainable params, device=cuda, epochs=30, patience=5, lr=0.001


[model] predict: 1427 samples, DO range [-1.52, 2.14] mg/L (scaled)
[data] load_gauge 2174: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.0, 'pH_sensor': 0.514, 'ec_sensor': 0.146, 'O2C_sensor': 0.108}
[data] split: train=12418, val=731, test=1461
[data] make_windows: 12384 windows, X=(12384, 21, 1), y=(12384, 14, 1), date range 1981-01-22 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 1), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1427 windows, X=(1427, 21, 1), y=(1427, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 12384 samples, X=(12384, 21, 1), y=(12384, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 1), y=(697, 14, 1)
[model] RiverDataset: 1427 samples, X=(1427, 21, 1), y=(1427, 14, 1)
[model] train_model: 100,929 trainable params, device=cuda, epochs=30, patience=5, lr=0.001


[model] predict: 1427 samples, DO range [-1.32, 1.89] mg/L (scaled)
[data] load_gauge 2174: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.0, 'pH_sensor': 0.514, 'ec_sensor': 0.146, 'O2C_sensor': 0.108}
[data] split: train=12418, val=731, test=1461
[data] make_windows: 12384 windows, X=(12384, 21, 2), y=(12384, 14, 1), date range 1981-01-22 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 2), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1427 windows, X=(1427, 21, 2), y=(1427, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 12384 samples, X=(12384, 21, 2), y=(12384, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 2), y=(697, 14, 1)
[model] RiverDataset: 1427 samples, X=(1427, 21, 2), y=(1427, 14, 1)
[model] train_model: 101,185 trainable params, device=cuda, epochs=30, patience=5, lr=0.001


[model] predict: 1427 samples, DO range [-1.43, 2.34] mg/L (scaled)
[data] load_gauge 2179: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.559, 'pH_sensor': 1.0, 'ec_sensor': 0.793, 'O2C_sensor': 1.0}
[data] split: train=12418, val=731, test=1461
[data] make_windows: 4242 windows, X=(4242, 21, 1), y=(4242, 14, 1), date range 2003-05-09 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 1), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1427 windows, X=(1427, 21, 1), y=(1427, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 4242 samples, X=(4242, 21, 1), y=(4242, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 1), y=(697, 14, 1)
[model] RiverDataset: 1427 samples, X=(1427, 21, 1), y=(1427, 14, 1)
[model] train_model: 100,929 trainable params, device=cuda, epochs=30, patience=5, lr=0.001


[model] predict: 1427 samples, DO range [-1.47, 1.81] mg/L (scaled)
[data] load_gauge 2179: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.559, 'pH_sensor': 1.0, 'ec_sensor': 0.793, 'O2C_sensor': 1.0}
[data] split: train=12418, val=731, test=1461
[data] make_windows: 4242 windows, X=(4242, 21, 2), y=(4242, 14, 1), date range 2003-05-09 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 2), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1427 windows, X=(1427, 21, 2), y=(1427, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 4242 samples, X=(4242, 21, 2), y=(4242, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 2), y=(697, 14, 1)
[model] RiverDataset: 1427 samples, X=(1427, 21, 2), y=(1427, 14, 1)
[model] train_model: 101,185 trainable params, device=cuda, epochs=30, patience=5, lr=0.001


[model] predict: 1427 samples, DO range [-1.46, 1.85] mg/L (scaled)
[data] load_gauge 2290: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.709, 'pH_sensor': 1.0, 'ec_sensor': 0.708, 'O2C_sensor': 1.0}
[data] split: train=12418, val=731, test=1461
[data] make_windows: 2064 windows, X=(2064, 21, 1), y=(2064, 14, 1), date range 2009-04-25 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 1), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1401 windows, X=(1401, 21, 1), y=(1401, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 2064 samples, X=(2064, 21, 1), y=(2064, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 1), y=(697, 14, 1)
[model] RiverDataset: 1401 samples, X=(1401, 21, 1), y=(1401, 14, 1)
[model] train_model: 100,929 trainable params, device=cuda, epochs=30, patience=5, lr=0.001


[model] predict: 1401 samples, DO range [-1.78, 2.81] mg/L (scaled)
[data] load_gauge 2290: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.709, 'pH_sensor': 1.0, 'ec_sensor': 0.708, 'O2C_sensor': 1.0}
[data] split: train=12418, val=731, test=1461
[data] make_windows: 2064 windows, X=(2064, 21, 2), y=(2064, 14, 1), date range 2009-04-25 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 2), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1401 windows, X=(1401, 21, 2), y=(1401, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 2064 samples, X=(2064, 21, 2), y=(2064, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 2), y=(697, 14, 1)
[model] RiverDataset: 1401 samples, X=(1401, 21, 2), y=(1401, 14, 1)
[model] train_model: 101,185 trainable params, device=cuda, epochs=30, patience=5, lr=0.001


[model] predict: 1401 samples, DO range [-1.58, 2.15] mg/L (scaled)
[data] load_gauge 2410: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.263, 'pH_sensor': 0.268, 'ec_sensor': 0.263, 'O2C_sensor': 0.265}
[data] split: train=12418, val=731, test=1461
[data] make_windows: 8533 windows, X=(8533, 21, 1), y=(8533, 14, 1), date range 1991-04-12 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 1), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1427 windows, X=(1427, 21, 1), y=(1427, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 8533 samples, X=(8533, 21, 1), y=(8533, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 1), y=(697, 14, 1)
[model] RiverDataset: 1427 samples, X=(1427, 21, 1), y=(1427, 14, 1)
[model] train_model: 100,929 trainable params, device=cuda, epochs=30, patience=5, lr=0.001


[model] predict: 1427 samples, DO range [-1.63, 2.24] mg/L (scaled)
[data] load_gauge 2410: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.263, 'pH_sensor': 0.268, 'ec_sensor': 0.263, 'O2C_sensor': 0.265}
[data] split: train=12418, val=731, test=1461
[data] make_windows: 8533 windows, X=(8533, 21, 2), y=(8533, 14, 1), date range 1991-04-12 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 2), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1427 windows, X=(1427, 21, 2), y=(1427, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 8533 samples, X=(8533, 21, 2), y=(8533, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 2), y=(697, 14, 1)
[model] RiverDataset: 1427 samples, X=(1427, 21, 2), y=(1427, 14, 1)
[model] train_model: 101,185 trainable params, device=cuda, epochs=30, patience=5, lr=0.001


[model] predict: 1427 samples, DO range [-1.64, 2.02] mg/L (scaled)
[data] load_gauge 2415: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.0, 'pH_sensor': 0.163, 'ec_sensor': 0.172, 'O2C_sensor': 0.171}
[data] split: train=12418, val=731, test=1461
[data] make_windows: 12384 windows, X=(12384, 21, 1), y=(12384, 14, 1), date range 1981-01-22 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 1), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1427 windows, X=(1427, 21, 1), y=(1427, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 12384 samples, X=(12384, 21, 1), y=(12384, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 1), y=(697, 14, 1)
[model] RiverDataset: 1427 samples, X=(1427, 21, 1), y=(1427, 14, 1)
[model] train_model: 100,929 trainable params, device=cuda, epochs=30, patience=5, lr=0.001


[model] predict: 1427 samples, DO range [-1.59, 1.75] mg/L (scaled)
[data] load_gauge 2415: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.0, 'pH_sensor': 0.163, 'ec_sensor': 0.172, 'O2C_sensor': 0.171}
[data] split: train=12418, val=731, test=1461
[data] make_windows: 12384 windows, X=(12384, 21, 2), y=(12384, 14, 1), date range 1981-01-22 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 2), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1427 windows, X=(1427, 21, 2), y=(1427, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 12384 samples, X=(12384, 21, 2), y=(12384, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 2), y=(697, 14, 1)
[model] RiverDataset: 1427 samples, X=(1427, 21, 2), y=(1427, 14, 1)
[model] train_model: 101,185 trainable params, device=cuda, epochs=30, patience=5, lr=0.001


[model] predict: 1427 samples, DO range [-1.60, 1.82] mg/L (scaled)
[data] load_gauge 2462: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.0, 'pH_sensor': 0.466, 'ec_sensor': 0.459, 'O2C_sensor': 0.543}
[data] split: train=12418, val=731, test=1461
[data] make_windows: 12384 windows, X=(12384, 21, 1), y=(12384, 14, 1), date range 1981-01-22 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 1), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1427 windows, X=(1427, 21, 1), y=(1427, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 12384 samples, X=(12384, 21, 1), y=(12384, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 1), y=(697, 14, 1)
[model] RiverDataset: 1427 samples, X=(1427, 21, 1), y=(1427, 14, 1)
[model] train_model: 100,929 trainable params, device=cuda, epochs=30, patience=5, lr=0.001


[model] predict: 1427 samples, DO range [-1.42, 1.56] mg/L (scaled)
[data] load_gauge 2462: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.0, 'pH_sensor': 0.466, 'ec_sensor': 0.459, 'O2C_sensor': 0.543}
[data] split: train=12418, val=731, test=1461
[data] make_windows: 12384 windows, X=(12384, 21, 2), y=(12384, 14, 1), date range 1981-01-22 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 2), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1427 windows, X=(1427, 21, 2), y=(1427, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 12384 samples, X=(12384, 21, 2), y=(12384, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 2), y=(697, 14, 1)
[model] RiverDataset: 1427 samples, X=(1427, 21, 2), y=(1427, 14, 1)
[model] train_model: 101,185 trainable params, device=cuda, epochs=30, patience=5, lr=0.001


[model] predict: 1427 samples, DO range [-1.50, 1.48] mg/L (scaled)
[data] load_gauge 2467: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.0, 'pH_sensor': 1.0, 'ec_sensor': 0.808, 'O2C_sensor': 1.0}
[data] split: train=12418, val=731, test=1461
[data] make_windows: 12384 windows, X=(12384, 21, 1), y=(12384, 14, 1), date range 1981-01-22 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 1), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1427 windows, X=(1427, 21, 1), y=(1427, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 12384 samples, X=(12384, 21, 1), y=(12384, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 1), y=(697, 14, 1)
[model] RiverDataset: 1427 samples, X=(1427, 21, 1), y=(1427, 14, 1)
[model] train_model: 100,929 trainable params, device=cuda, epochs=30, patience=5, lr=0.001


[model] predict: 1427 samples, DO range [-1.57, 1.84] mg/L (scaled)
[data] load_gauge 2467: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.0, 'pH_sensor': 1.0, 'ec_sensor': 0.808, 'O2C_sensor': 1.0}
[data] split: train=12418, val=731, test=1461
[data] make_windows: 12384 windows, X=(12384, 21, 2), y=(12384, 14, 1), date range 1981-01-22 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 2), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1427 windows, X=(1427, 21, 2), y=(1427, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 12384 samples, X=(12384, 21, 2), y=(12384, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 2), y=(697, 14, 1)
[model] RiverDataset: 1427 samples, X=(1427, 21, 2), y=(1427, 14, 1)
[model] train_model: 101,185 trainable params, device=cuda, epochs=30, patience=5, lr=0.001


[model] predict: 1427 samples, DO range [-1.60, 1.87] mg/L (scaled)
[data] load_gauge 2613: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.35, 'pH_sensor': 0.023, 'ec_sensor': 0.008, 'O2C_sensor': 0.016}
[data] split: train=12418, val=731, test=1461
[data] make_windows: 7292 windows, X=(7292, 21, 1), y=(7292, 14, 1), date range 1995-01-01 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 1), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1427 windows, X=(1427, 21, 1), y=(1427, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 7292 samples, X=(7292, 21, 1), y=(7292, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 1), y=(697, 14, 1)
[model] RiverDataset: 1427 samples, X=(1427, 21, 1), y=(1427, 14, 1)
[model] train_model: 100,929 trainable params, device=cuda, epochs=30, patience=5, lr=0.001


[model] predict: 1427 samples, DO range [-1.57, 2.09] mg/L (scaled)
[data] load_gauge 2613: 14610 rows, 1981-01-01 → 2020-12-31, cols=['temp_sensor', 'pH_sensor', 'ec_sensor', 'O2C_sensor']
[data] preprocess: 14610 days, NaN%={'temp_sensor': 0.35, 'pH_sensor': 0.023, 'ec_sensor': 0.008, 'O2C_sensor': 0.016}
[data] split: train=12418, val=731, test=1461
[data] make_windows: 7292 windows, X=(7292, 21, 2), y=(7292, 14, 1), date range 1995-01-01 → 2014-12-18
[data] make_windows: 697 windows, X=(697, 21, 2), y=(697, 14, 1), date range 2015-01-22 → 2016-12-18
[data] make_windows: 1427 windows, X=(1427, 21, 2), y=(1427, 14, 1), date range 2017-01-22 → 2020-12-18
[model] RiverDataset: 7292 samples, X=(7292, 21, 2), y=(7292, 14, 1)
[model] RiverDataset: 697 samples, X=(697, 21, 2), y=(697, 14, 1)
[model] RiverDataset: 1427 samples, X=(1427, 21, 2), y=(1427, 14, 1)
[model] train_model: 101,185 trainable params, device=cuda, epochs=30, patience=5, lr=0.001


[model] predict: 1427 samples, DO range [-1.59, 2.22] mg/L (scaled)
Multi-gauge results (14 gauges):
  Temp-only: mean RMSE=1.451°C  NSE=0.881
  Temp+EC:   mean RMSE=1.414°C  NSE=0.888


Saved: temp_ec_results.csv


## 4. Note on Actual Precipitation Integration


### Framework for CAMELS-CH Precipitation Forcing

When the CAMELS-CH base meteorological forcing (Höge et al. 2023,
https://essd.copernicus.org/articles/15/5755/2023/) is available,
precipitation can be added directly as an input feature:

```python
# FEATURES_TEMP_PRECIP = ['temp_sensor', 'precip_mm']  # replace ec_sensor
```

The CAMELS-CH base dataset provides daily catchment-averaged precipitation
for all 331 Swiss catchments (including our 86 CAMELS-CH-Chem gauges)
via the same gauge_id linkage used for static attributes.

Expected improvement: precipitation-informed temperature forecasts may
reduce RMSE at high-elevation gauges where snowmelt drives temperature
variability (see `attribute_rmse_correlation.csv` — elevation proxy correlates
with temperature RMSE).


In [5]:
# Build comparison summary from df_mg
if 'df_mg' not in dir() or df_mg.empty:
    print("No multi-gauge results available")
else:
    # Pivot to get base and ec side by side
    pivot = df_mg.pivot_table(index='gauge_id', columns='model', 
                               values=['rmse_temp', 'nse_temp'])
    pivot.columns = ['_'.join(c) for c in pivot.columns]
    valid_multi = pivot.dropna()

    summary = pd.DataFrame([
        {
            'model':        'Temp-only (this nb)',
            'mean_rmse':    valid_multi['rmse_temp_base'].mean(),
            'mean_nse':     valid_multi['nse_temp_base'].mean(),
            'n_gauges':     len(valid_multi),
        },
        {
            'model':        'Temp + EC (this nb)',
            'mean_rmse':    valid_multi['rmse_temp_ec'].mean(),
            'mean_nse':     valid_multi['nse_temp_ec'].mean(),
            'n_gauges':     len(valid_multi),
        },
        {
            'model':        'Zero-shot (nb04b)',
            'mean_rmse':    2.597,
            'mean_nse':     0.727,
            'n_gauges':     15,
        },
        {
            'model':        'EA-LSTM (nb04b)',
            'mean_rmse':    1.721,
            'mean_nse':     0.862,
            'n_gauges':     15,
        },
    ])

    print("Temperature Forecast Comparison")
    print("=" * 60)
    print(summary.to_string(index=False))
    print()
    delta = valid_multi['rmse_temp_base'].mean() - valid_multi['rmse_temp_ec'].mean()
    print(f"EC improvement: {delta:+.3f}°C ({delta/valid_multi['rmse_temp_base'].mean()*100:+.1f}%)")

    out_summary = RESULTS_DIR / 'temp_forecast_comparison.csv'
    summary.to_csv(out_summary, index=False)
    print(f"Saved: {out_summary}")


Temperature Forecast Comparison
              model  mean_rmse  mean_nse  n_gauges
Temp-only (this nb)   1.451036  0.880571        14
Temp + EC (this nb)   1.413700  0.887929        14
  Zero-shot (nb04b)   2.597000  0.727000        15
    EA-LSTM (nb04b)   1.721000  0.862000        15

EC improvement: +0.037°C (+2.6%)
Saved: /storage/homefs/tn20y076/AareML/notebooks/../results/temp_forecast_comparison.csv
